File: app/src/main/java/f4/hg/wrappers/ZeroInsetsFrameLayout.java
package f4.hg.wrappers;

import android.content.Context;
import android.graphics.Rect;
import android.os.Build;
import android.support.annotation.NonNull;
import android.util.AttributeSet;
import android.view.WindowInsets;
import android.widget.FrameLayout;

/*
 * Huge thanks and credits to Hogun (https://stackoverflow.com/a/41448403) for the class.
 */

public final class ZeroInsetsFrameLayout extends FrameLayout {
    private int[] mInsets = new int[4];

    public ZeroInsetsFrameLayout(Context context) {
        super(context);
    }

    public ZeroInsetsFrameLayout(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public ZeroInsetsFrameLayout(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
    }

    public final int[] getInsets() {
        return mInsets;
    }

    @Override
    public WindowInsets computeSystemWindowInsets(WindowInsets in, Rect outLocalInsets) {
        outLocalInsets.left = 0;
        outLocalInsets.top = 0;
        outLocalInsets.right = 0;

        return super.computeSystemWindowInsets(in, outLocalInsets);
    }

    @Override
    protected final boolean fitSystemWindows(@NonNull Rect insets) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            // Intentionally do not modify the bottom inset. For some reason,
            // if the bottom inset is modified, window resizing stops working.
            // TODO: Figure out why.

            mInsets[0] = insets.left;
            mInsets[1] = insets.top;
            mInsets[2] = insets.right;

            insets.left = 0;
            insets.top = 0;
            insets.right = 0;
        }

        return super.fitSystemWindows(insets);
    }
}

File: app/src/main/java/mono/hg/views/ZeroInsetsFrameLayout.kt
package mono.hg.views

import android.content.Context
import android.graphics.Rect
import android.os.Build
import android.util.AttributeSet
import android.view.WindowInsets
import android.widget.FrameLayout

/**
 * FrameLayout workaround for transparency issue in Android 4.4 and above.
 *
 * Huge thanks and credits to Hogun (https://stackoverflow.com/a/41448403) for the class.
 */
class ZeroInsetsFrameLayout : FrameLayout {
    private val insets = IntArray(4)

    constructor(context: Context?) : super(context !!) {
        // Unused. Used only in XML initialisation.
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context !!, attrs) {
        // Unused. Used only in XML initialisation.
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context !!,
        attrs,
        defStyle
    ) {
        // Unused. Used only in XML initialisation.
    }

    override fun computeSystemWindowInsets(`in`: WindowInsets, outLocalInsets: Rect): WindowInsets {
        outLocalInsets.left = 0
        outLocalInsets.top = 0
        outLocalInsets.right = 0
        return super.computeSystemWindowInsets(`in`, outLocalInsets)
    }

    override fun fitSystemWindows(windowInsets: Rect): Boolean {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            // Intentionally do not modify the bottom inset. For some reason,
            // if the bottom inset is modified, window resizing stops working.
            // TODO: Figure out why.
            insets[0] = windowInsets.left
            insets[1] = windowInsets.top
            insets[2] = windowInsets.right
            windowInsets.left = 0
            windowInsets.top = 0
            windowInsets.right = 0
        }
        return super.fitSystemWindows(windowInsets)
    }
}